pkg/server/controllers.go

```diff
@@ -600,21 +682,20 @@ func (s *Server) installAPIBindingController(ctx context.Context, config *rest.C
 		return err
 	}
 
-	c, err := apibinding.NewController(
-		crdClusterClient,
-		kcpClusterClient,
-		s.KcpSharedInformerFactory.Apis().V1alpha1().APIBindings(),
-		s.KcpSharedInformerFactory.Apis().V1alpha1().APIExports(),
-		s.KcpSharedInformerFactory.Apis().V1alpha1().APIResourceSchemas(),
-		s.KcpSharedInformerFactory.Apis().V1alpha1().APIConversions(),
-		s.CacheKcpSharedInformerFactory.Apis().V1alpha1().APIExports(),
-		s.CacheKcpSharedInformerFactory.Apis().V1alpha1().APIResourceSchemas(),
-		s.CacheKcpSharedInformerFactory.Apis().V1alpha1().APIConversions(),
-		s.ApiExtensionsSharedInformerFactory.Apiextensions().V1().CustomResourceDefinitions(),
-	)
-	if err != nil {
-		return err
-	}
+	// APIBinding indexers
+	indexers.AddIfNotPresentOrDie(s.KcpSharedInformerFactory.Apis().V1alpha1().APIBindings().Informer().GetIndexer(), cache.Indexers{
+		indexers.APIBindingsByAPIExport: indexers.IndexAPIBindingByAPIExport,
+	})
+
+	// APIExport indexers
+	indexers.AddIfNotPresentOrDie(s.KcpSharedInformerFactory.Apis().V1alpha1().APIExports().Informer().GetIndexer(), cache.Indexers{
+		indexers.ByLogicalClusterPathAndName: indexers.IndexByLogicalClusterPathAndName,
+		indexAPIExportsByAPIResourceSchema:   indexAPIExportsByAPIResourceSchemasFunc,
+	})
+	indexers.AddIfNotPresentOrDie(s.CacheKcpSharedInformerFactory.Apis().V1alpha1().APIExports().Informer().GetIndexer(), cache.Indexers{
+		indexers.ByLogicalClusterPathAndName: indexers.IndexByLogicalClusterPathAndName,
+		indexAPIExportsByAPIResourceSchema:   indexAPIExportsByAPIResourceSchemasFunc,
+	})
```

Comment:
What is the reason for moving the indexer setup out of the controller creation to here (this happens in other parts of the code, so this is more of a placeholder reference for all those changes)?

Reply:
The core reason we moved registration of indexers outside the `controller.NewFooController` is because with this change, all of the `controller.NewFooController` calls are inside the `Runner` function which is called everytime a KCP pod becomes a leader.

`indexers.AddIfNotPresentOrDie` is supposed to be run only once in the lifecycle of a KCP pod because otherwise it will panic.

The mental model is that:
- installFooController outside of Runner has steps that can or needed to run one-time only in the lifecycle of the KCP pod
- Runner has all the steps that can be run multiple times in the runtime of a Pod.

Note: The same KCP pod can somehow lose leader-election and get elected again in due course of time. The changes in this PR enables that scenario.

Reply:
@embik this is a little bit complicated to wrap at one go. Maybe we can explain over the community call later.

Reply:
@palnabarun I first understood `indexers.AddIfNotPresentOrDie` to be idem-potent since it filters out already present indexers (that's why I asked), but my feeling is the problem isn't _that_ part of the logic, it's that calling `indexer.AddIndexers` even with an empty list of indexers (because `indexers.AddIfNotPresentOrDie` filtered out all existing indexers) will panic when the store is already started -- do I get this right? Or is the problem _that_ it's called with an empty list of indexers (because they all already exist)?

I won't be on the next community call unfortunately, but I don't mind anyone else reviewing this and discussing it there either if that works better.